cranelift/codegen/src/machinst/lower.rs

```diff
@@ -172,17 +163,22 @@ pub trait LowerCtx {
     fn ensure_in_vreg(&mut self, reg: Reg, ty: Type) -> Reg;
 }
 
-/// A representation of all of the ways in which an instruction input is
-/// available: as a producing instruction (in the same color-partition), as a
-/// constant, and/or in an existing register. See [LowerCtx::get_input] for more
-/// details.
+/// A representation of all of the ways in which a value is available, aside
+/// from as a direct register.
+///
+/// - An instruction, if it would be allowed to occur at the current location
+///   instead (see [LowerCtx::maybe_get_input_as_source_or_const()] for more
+///   details).
+///
+/// - A constant, if the value is known to be a constant.
 #[derive(Clone, Copy, Debug)]
 pub struct LowerInput {
-    /// The value is live in a register. This option is always available. Call
-    /// [LowerCtx::use_input_reg()] if the register is used.
-    pub reg: Reg,
-    /// An instruction produces this value; the instruction's result index that
-    /// produces this value is given.
+    /// An instruction produces this value (as the given output), and its
+    /// computation (and side-effect if applicable) could occur at the
+    /// current instruction's location instead.
+    ///
+    /// If this instruction's operation is merged into the current instruction,
+    /// the backend must call [LowerCtx::sink_inst()].
     pub inst: Option<(Inst, usize)>,
     /// The value is a known constant.
     pub constant: Option<u64>,
```

Comment:
Can `inst` and `constant` ever *both* be `Some`? If not, then I think it makes more sense to define this as an `enum NonRegInput` and then methods that create this would return `Option<NonRegInput>` and use `None` instead of returning a struct that has empty components.

Reply:
The `inst` and `constant` fields can actually both be `Some`, and in fact some of the matching code relies on this -- it matches on the opcode implied by `inst` where one of the options is `Iconst`. In other words when we know a constant value for the input it's because we know its source instruction, and that source instruction is iconst/fconst/etc. Otherwise I agree this would have been a good simplification!